x/group/keeper/msg_server.go

```diff
@@ -810,6 +810,105 @@ func (k Keeper) Exec(goCtx context.Context, req *group.MsgExec) (*group.MsgExecR
 	return res, nil
 }
 
+// LeaveGroup implements the MsgServer/LeaveGroup method.
+func (k Keeper) LeaveGroup(goCtx context.Context, req *group.MsgLeaveGroup) (*group.MsgLeaveGroupResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	_, err := sdk.AccAddressFromBech32(req.Address)
+	if err != nil {
+		return nil, err
+	}
+
+	groupInfo, err := k.getGroupInfo(ctx, req.GroupId)
+	if err != nil {
+		return nil, sdkerrors.Wrap(err, "group")
+	}
+
+	groupWeight, err := math.NewNonNegativeDecFromString(groupInfo.TotalWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	gm, err := k.getGroupMember(ctx, &group.GroupMember{
+		GroupId: req.GroupId,
+		Member:  &group.Member{Address: req.Address},
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	memberWeight, err := math.NewNonNegativeDecFromString(gm.Member.Weight)
+	if err != nil {
+		return nil, err
+	}
+
+	updatedWeight, err := math.SubNonNegative(groupWeight, memberWeight)
+	if err != nil {
+		return nil, err
+	}
+
+	iter, err := k.getGroupPoliciesByGroup(ctx, req.GroupId, nil)
+	if err != nil {
+		return nil, err
+	}
+	defer iter.Close()
+
+	// update group weight and increment group version
+	groupInfo.TotalWeight = updatedWeight.String()
+	groupInfo.Version++
+
+	var groupPolicy group.GroupPolicyInfo
+	for {
+		ctx.GasMeter().ConsumeGas(gasCostPerIteration, "group policy")
+		if _, err := iter.LoadNext(&groupPolicy); err != nil {
+			if errors.ErrORMIteratorDone.Is(err) {
+				break
+			}
+
+			return nil, err
+		}
+
+		policyI := groupPolicy.GetDecisionPolicy()
+		if policyI == nil {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("expected %T, got %T", (group.DecisionPolicy)(nil), groupPolicy.DecisionPolicy.GetCachedValue())
+		}
```

Comment:
Are we not updating the `threshold` in `ThresholdDecisionPolicy`?

Reply:
First, we should have consistent behaviors whether a group member is removed through the existing `MsgUpdateGroupMembers` or through `MsgLeaveGroup`
currently in `MsgUpdateGroupMembers` if we remove a member that invalidates a threshold decision policy, we don't do anything, but then proposals won't be able to be created.
As discussed here: https://github.com/cosmos/cosmos-sdk/issues/10945, the other options that seem better from a UX standpoint are indeed:
1. either to update the threshold
2. or to prevent the member from leaving the group
The drawback of 1. is that automatic threshold updates might not be desired in certain use cases...

As far as I can remember, we haven't really come to a decision on this (no feedback yet on #10945).
I guess in the meantime, we could just keep the current behavior (ie do nothing) in this PR and then address any desired change as part of a follow-up PR that would close #10945

cc/ @aaronc @amaurym @cmwaters

Reply:
Just realized that the issue now with doing nothing on member leave is that it might make current proposals impossible to pass. In the initial x/group version, proposals got unvalidated as soon as there was some member update. But now that tally is done at the end, that might be an issue so we should consider implementing 1 or 2.

Reply:
@aleem1314 let's just do nothing as part of this PR (so please remove the policy validation in `MsgLeaveGroup`) and we'll implement what @cmwaters proposed in https://github.com/cosmos/cosmos-sdk/issues/10945#issuecomment-1057943824 as part of a separate PR